File: backends/drogon/src/llm_client.cc
#include "llm_client.h"

#include <drogon/HttpClient.h>
#include <trantor/utils/Logger.h>

#include <chrono>

namespace
{
std::string compactJson(const Json::Value &v)
{
    Json::StreamWriterBuilder builder;
    builder["indentation"] = "";
    return Json::writeString(builder, v);
}

// base_url 拆为 origin（scheme://host:port，给 HttpClient）+ path 前缀。
struct BaseUrl
{
    std::string origin;
    std::string pathPrefix;
};

BaseUrl splitBaseUrl(const std::string &raw)
{
    std::string url = raw;
    // trim + 去尾部斜杠
    while (!url.empty() && (url.back() == '/' || url.back() == ' '))
        url.pop_back();
    if (url.empty())
        url = "https://api.openai.com/v1";
    const auto schemeEnd = url.find("://");
    if (schemeEnd == std::string::npos)
        return {"https://" + url, ""};
    const auto pathStart = url.find('/', schemeEnd + 3);
    if (pathStart == std::string::npos)
        return {url, ""};
    return {url.substr(0, pathStart), url.substr(pathStart)};
}

std::string extractTokens(const Json::Value &body)
{
    if (!body.isMember("usage") || !body["usage"].isObject())
        return "n/a";
    const auto &usage = body["usage"];
    int64_t total = 0;
    if (LlmClient::valueAsInt(usage["total_tokens"], total) && total > 0)
        return std::to_string(total);
    int64_t input = 0, output = 0;
    LlmClient::valueAsInt(usage["input_tokens"], input);
    LlmClient::valueAsInt(usage["output_tokens"], output);
    if (input + output > 0)
        return std::to_string(input + output);
    return "n/a";
}

std::string extractChatText(const Json::Value &body)
{
    const auto &choices = body["choices"];
    if (choices.isArray() && !choices.empty())
    {
        const auto &content = choices[0]["message"]["content"];
        if (content.isString() && !content.asString().empty())
            return content.asString();
    }
    throw LlmError("LLM chat response missing content");
}

std::string extractResponsesText(const Json::Value &body)
{
    if (body["output_text"].isString() && !body["output_text"].asString().empty())
        return body["output_text"].asString();
    if (body["output"].isArray())
    {
        for (const auto &item : body["output"])
        {
            if (!item["content"].isArray())
                continue;
            for (const auto &entry : item["content"])
                if (entry["text"].isString() && !entry["text"].asString().empty())
                    return entry["text"].asString();
        }
    }
    throw LlmError("LLM response did not contain output text");
}
}  // namespace

drogon::Task<Json::Value> LlmClient::generateCapsuleSuggestion(std::string prompt)
{
    co_return co_await generateStructuredJson(std::move(prompt), suggestionSpec());
}

drogon::Task<Json::Value> LlmClient::generateCapsuleRecommendations(std::string prompt)
{
    co_return co_await generateStructuredJson(std::move(prompt), recommendationSpec());
}

drogon::Task<Json::Value> LlmClient::generateStructuredJson(std::string prompt,
                                                            const SchemaSpec &spec)
{
    if (!config_.enabled || config_.apiKey.empty())
        throw LlmError("LLM is disabled or missing API key");
    if (config_.apiStyle == "responses")
        co_return co_await generateWithResponses(std::move(prompt), spec);
    if (config_.apiStyle == "auto")
    {
        // C++ 协程不允许在 catch 块里 co_await：先记失败，try 外再回落 chat。
        bool fallbackToChat = false;
        Json::Value result;
        try
        {
            result = co_await generateWithResponses(prompt, spec);
        }
        catch (const LlmError &e)
        {
            LOG_INFO << "Responses API unavailable (" << e.what()
                     << "); falling back to chat completions";
            fallbackToChat = true;
        }
        if (!fallbackToChat)
            co_return result;
    }
    co_return co_await generateWithChat(std::move(prompt), spec, true);
}

drogon::Task<Json::Value> LlmClient::generateWithResponses(std::string prompt,
                                                           const SchemaSpec &spec)
{
    Json::Value format(Json::objectValue);
    format["type"] = "json_schema";
    format["name"] = spec.schemaName;
    format["strict"] = true;
    format["schema"] = spec.schema;
    Json::Value text(Json::objectValue);
    text["format"] = format;

    Json::Value payload(Json::objectValue);
    payload["model"] = config_.model;
    payload["input"] = prompt;
    payload["max_output_tokens"] = spec.maxOutputTokens;
    payload["text"] = text;

    const Json::Value body = co_await postJson("/responses", payload);
    co_return parseJsonObject(extractResponsesText(body));
}

drogon::Task<Json::Value> LlmClient::generateWithChat(std::string prompt,
                                                      const SchemaSpec &spec,
                                                      bool disableThinking)
{
    const auto buildPayload = [&](bool withThinking) {
        Json::Value messages(Json::arrayValue);
        Json::Value sys(Json::objectValue);
        sys["role"] = "system";
        sys["content"] = spec.systemPrompt;
        Json::Value usr(Json::objectValue);
        usr["role"] = "user";
        usr["content"] = prompt;
        messages.append(sys);
        messages.append(usr);

        Json::Value payload(Json::objectValue);
        payload["model"] = config_.model;
        payload["messages"] = messages;
        payload["max_tokens"] = spec.maxTokens;
        if (withThinking)
        {
            Json::Value thinking(Json::objectValue);
            thinking["type"] = "disabled";
            payload["thinking"] = thinking;
        }
        return payload;
    };

    // 某些网关不认 thinking 字段：400 时去掉重试一次（co_await 须在 catch 外）。
    bool retryWithoutThinking = false;
    Json::Value body;
    try
    {
        body = co_await postJson("/chat/completions", buildPayload(disableThinking));
    }
    catch (const LlmError &e)
    {
        if (e.status == 400 && disableThinking)
            retryWithoutThinking = true;
        else
            throw;
    }
    if (retryWithoutThinking)
        body = co_await postJson("/chat/completions", buildPayload(false));
    co_return parseJsonObject(extractChatText(body));
}

// 向 path POST JSON；瞬时网络/TLS 错误按配置重试，HTTP 4xx/5xx 与坏 JSON 不重试。
drogon::Task<Json::Value> LlmClient::postJson(std::string path, Json::Value payload)
{
    const BaseUrl base = splitBaseUrl(config_.baseUrl);
    const std::string url = base.origin + base.pathPrefix + path;
    const std::string bodyStr = compactJson(payload);
    const int attempts = std::max(1, config_.maxRetries + 1);
    std::string lastError = "LLM request failed";

    for (int attempt = 1; attempt <= attempts; ++attempt)
    {
        LOG_INFO << "LLM request  model=" << config_.model << " url=" << url
                 << " attempt=" << attempt << "/" << attempts;
        const auto start = std::chrono::steady_clock::now();
        const auto elapsedMs = [&start] {
            return std::chrono::duration_cast<std::chrono::milliseconds>(
                       std::chrono::steady_clock::now() - start)
                .count();
        };
        try
        {
            auto client = drogon::HttpClient::newHttpClient(base.origin);
            auto req = drogon::HttpRequest::newHttpRequest();
            req->setMethod(drogon::Post);
            req->setPath(base.pathPrefix + path);
            req->addHeader("Authorization", "Bearer " + config_.apiKey);
            req->addHeader("Accept", "application/json");
            req->addHeader("User-Agent", config_.userAgent);
            req->setContentTypeCode(drogon::CT_APPLICATION_JSON);
            req->setBody(bodyStr);

            const auto resp = co_await client->sendRequestCoro(
                req, static_cast<double>(config_.timeoutMs) / 1000.0);
            const int status = static_cast<int>(resp->statusCode());
            if (status < 200 || status >= 300)
            {
                LOG_WARN << "LLM error    model=" << config_.model
                         << " elapsed_ms=" << elapsedMs() << " status=" << status;
                const std::string text(resp->body());
                throw LlmError("HTTP " + std::to_string(status) + ": " + text.substr(0, 500),
                               status);
            }
            Json::Value parsed;
            Json::CharReaderBuilder builder;
            std::string errs;
            const std::unique_ptr<Json::CharReader> reader(builder.newCharReader());
            const auto body = resp->body();
            if (!reader->parse(body.data(), body.data() + body.size(), &parsed, &errs) ||
                !parsed.isObject())
            {
                LOG_WARN << "LLM error    model=" << config_.model
                         << " elapsed_ms=" << elapsedMs() << " error=invalid-json";
                throw LlmError("LLM response was not valid JSON");
            }
            LOG_INFO << "LLM response model=" << config_.model
                     << " elapsed_ms=" << elapsedMs() << " tokens=" << extractTokens(parsed);
            co_return parsed;
        }
        catch (const LlmError &)
        {
            throw;  // HTTP 状态错误/坏 JSON 不重试
        }
        catch (const std::exception &e)
        {
            const bool willRetry = attempt < attempts;
            LOG_WARN << "LLM error    model=" << config_.model
                     << " elapsed_ms=" << elapsedMs() << " error=" << e.what()
                     << (willRetry ? " (will retry)" : "");
            lastError = e.what();
        }
        // 退避在 catch 外（协程限制），仅瞬时传输错误会走到这里。
        if (attempt < attempts)
            co_await drogon::sleepCoro(
                trantor::EventLoop::getEventLoopOfCurrentThread(),
                static_cast<double>(config_.retryBackoffMs * attempt) / 1000.0);
    }
    throw LlmError(lastError);
}

Json::Value LlmClient::parseJsonObject(const std::string &raw)
{
    std::string text = raw;
    // trim
    while (!text.empty() && (text.front() == ' ' || text.front() == '\n' ||
                             text.front() == '\r' || text.front() == '\t'))
        text.erase(text.begin());
    while (!text.empty() && (text.back() == ' ' || text.back() == '\n' ||
                             text.back() == '\r' || text.back() == '\t'))
        text.pop_back();
    // 剥 ``` 围栏
    if (text.rfind("```", 0) == 0)
    {
        auto firstNewline = text.find('\n');
        if (firstNewline != std::string::npos)
            text = text.substr(firstNewline + 1);
        if (text.size() >= 3 && text.compare(text.size() - 3, 3, "```") == 0)
            text = text.substr(0, text.size() - 3);
    }

    const auto tryParse = [](const std::string &s, Json::Value &out) {
        Json::CharReaderBuilder builder;
        std::string errs;
        const std::unique_ptr<Json::CharReader> reader(builder.newCharReader());
        return reader->parse(s.data(), s.data() + s.size(), &out, &errs) && out.isObject();
    };

    Json::Value parsed;
    if (tryParse(text, parsed))
        return parsed;
    const auto start = text.find('{');
    const auto end = text.rfind('}');
    if (start == std::string::npos || end == std::string::npos || start >= end)
        throw LlmError("LLM output was not valid JSON");
    if (!tryParse(text.substr(start, end - start + 1), parsed))
        throw LlmError("LLM output was not valid JSON");
    return parsed;
}

bool LlmClient::valueAsInt(const Json::Value &v, int64_t &out)
{
    if (v.isIntegral())
    {
        out = v.asInt64();
        return true;
    }
    if (v.isDouble())
    {
        out = static_cast<int64_t>(v.asDouble());
        return true;
    }
    return false;
}

// ── 结构化输出 schema ───────────────────────────────────────────────────────

const LlmClient::SchemaSpec &LlmClient::suggestionSpec()
{
    static const SchemaSpec spec = [] {
        Json::Value schema(Json::objectValue);
        schema["type"] = "object";
        schema["additionalProperties"] = false;
        Json::Value required(Json::arrayValue);
        required.append("title");
        required.append("content");
        required.append("openInDays");
        schema["required"] = required;
        Json::Value props(Json::objectValue);
        Json::Value strType(Json::objectValue);
        strType["type"] = "string";
        props["title"] = strType;
        props["content"] = strType;
        Json::Value days(Json::objectValue);
        days["type"] = "integer";
        days["minimum"] = 1;
        days["maximum"] = 3650;
        props["openInDays"] = days;
        schema["properties"] = props;
        return SchemaSpec{
            "capsule_suggestion", schema,
            "你只返回严格 JSON 对象，不要 Markdown、代码块或解释。"
            "JSON 必须包含字符串字段 title、content 和整数字段 openInDays。"
            "若用户已给出标题，title 可原样回填。",
            900, 900};
    }();
    return spec;
}

const LlmClient::SchemaSpec &LlmClient::recommendationSpec()
{
    static const SchemaSpec spec = [] {
        Json::Value itemSchema(Json::objectValue);
        itemSchema["type"] = "object";
        itemSchema["additionalProperties"] = false;
        Json::Value itemRequired(Json::arrayValue);
        itemRequired.append("title");
        itemRequired.append("hint");
        itemRequired.append("openInDays");
        itemSchema["required"] = itemRequired;
        Json::Value itemProps(Json::objectValue);
        Json::Value strType(Json::objectValue);
        strType["type"] = "string";
        itemProps["title"] = strType;
        itemProps["hint"] = strType;
        Json::Value days(Json::objectValue);
        days["type"] = "integer";
        days["minimum"] = 1;
        days["maximum"] = 3650;
        itemProps["openInDays"] = days;
        itemSchema["properties"] = itemProps;

        Json::Value items(Json::objectValue);
        items["type"] = "array";
        items["minItems"] = 3;
        items["maxItems"] = 8;
        items["items"] = itemSchema;

        Json::Value schema(Json::objectValue);
        schema["type"] = "object";
        schema["additionalProperties"] = false;
        Json::Value required(Json::arrayValue);
        required.append("items");
        schema["required"] = required;
        Json::Value props(Json::objectValue);
        props["items"] = items;
        schema["properties"] = props;
        return SchemaSpec{
            "capsule_recommendations", schema,
            "你只返回严格 JSON 对象，不要 Markdown、代码块或解释。"
            "JSON 必须包含数组字段 items，每项含字符串字段 title、hint 和整数字段 openInDays。",
            900, 900};
    }();
    return spec;
}

File: backends/vapor/server/Sources/App/Services/LlmClient.swift
import Foundation
import Vapor

/// 结构化 JSON 生成的 LLM 客户端。日志规范、网关重试、CF 1010 等坑见 docs/dev-notes.md §3。
/// 契约测试默认 LLM_ENABLED=false，会在此立即抛异常，由上层走本地兜底；HTTP 路径不在测试链上。
struct LlmClient: Sendable {
    struct LlmError: Error {
        let message: String
        var status: Int = 0
    }

    struct SchemaSpec: Sendable {
        let schemaName: String
        let schema: JSON
        let systemPrompt: String
        let maxOutputTokens: Int
        let maxTokens: Int
    }

    let config: LlmConfig
    let client: any Client
    let logger: Logger

    func generateCapsuleSuggestion(prompt: String) async throws -> JSON {
        try await generateStructuredJson(prompt: prompt, spec: Self.suggestionSpec)
    }

    func generateCapsuleRecommendations(prompt: String) async throws -> JSON {
        try await generateStructuredJson(prompt: prompt, spec: Self.recommendationSpec)
    }

    private func generateStructuredJson(prompt: String, spec: SchemaSpec) async throws -> JSON {
        guard config.enabled, !config.apiKey.isEmpty else {
            throw LlmError(message: "LLM is disabled or missing API key")
        }
        switch config.apiStyle {
        case "responses":
            return try await generateWithResponses(prompt: prompt, spec: spec)
        case "auto":
            do {
                return try await generateWithResponses(prompt: prompt, spec: spec)
            } catch let e as LlmError {
                logger.info("Responses API unavailable (\(e.message)); falling back to chat completions")
                return try await generateWithChat(prompt: prompt, spec: spec, disableThinking: true)
            }
        default:
            return try await generateWithChat(prompt: prompt, spec: spec, disableThinking: true)
        }
    }

    private func generateWithResponses(prompt: String, spec: SchemaSpec) async throws -> JSON {
        let payload: JSON = .object([
            "model": .string(config.model),
            "input": .string(prompt),
            "max_output_tokens": .int(spec.maxOutputTokens),
            "text": .object([
                "format": .object([
                    "type": .string("json_schema"),
                    "name": .string(spec.schemaName),
                    "strict": .bool(true),
                    "schema": spec.schema,
                ]),
            ]),
        ])
        let body = try await postJson(url: responsesUrl, payload: payload)
        return try parseJsonObject(extractResponsesText(body))
    }

    private func generateWithChat(prompt: String, spec: SchemaSpec, disableThinking: Bool) async throws -> JSON {
        func payload(withThinking: Bool) -> JSON {
            var obj: [String: JSON] = [
                "model": .string(config.model),
                "messages": .array([
                    .object(["role": .string("system"), "content": .string(spec.systemPrompt)]),
                    .object(["role": .string("user"), "content": .string(prompt)]),
                ]),
                "max_tokens": .int(spec.maxTokens),
            ]
            if withThinking { obj["thinking"] = .object(["type": .string("disabled")]) }
            return .object(obj)
        }
        do {
            let body = try await postJson(url: chatUrl, payload: payload(withThinking: disableThinking))
            return try parseJsonObject(extractChatText(body))
        } catch let e as LlmError where e.status == 400 && disableThinking {
            // 某些网关不认 thinking 字段，去掉重试一次。
            let body = try await postJson(url: chatUrl, payload: payload(withThinking: false))
            return try parseJsonObject(extractChatText(body))
        }
    }

    /// 向 url POST JSON；瞬时网络/TLS 错误按配置重试，HTTP 4xx/5xx 与坏 JSON 不重试。
    private func postJson(url: String, payload: JSON) async throws -> JSON {
        let bodyData = try JSONEncoder().encode(payload)
        let attempts = max(1, config.maxRetries + 1)
        var lastError: Error?

        for attempt in 1...attempts {
            logger.info("LLM request  model=\(config.model) url=\(url) attempt=\(attempt)/\(attempts)")
            let start = Date()
            do {
                var headers = HTTPHeaders()
                headers.add(name: .authorization, value: "Bearer \(config.apiKey)")
                headers.add(name: .contentType, value: "application/json")
                headers.add(name: .accept, value: "application/json")
                headers.add(name: .userAgent, value: config.userAgent)
                let response = try await client.post(URI(string: url), headers: headers) { req in
                    req.body = ByteBuffer(data: bodyData)
                }
                let elapsed = Int(Date().timeIntervalSince(start) * 1000)
                let status = Int(response.status.code)
                guard (200..<300).contains(status) else {
                    logger.warning("LLM error    model=\(config.model) elapsed_ms=\(elapsed) status=\(status)")
                    let bodyText = response.body.map { String(buffer: $0) } ?? ""
                    throw LlmError(message: "HTTP \(status): \(bodyText.prefix(500))", status: status)
                }
                guard let buffer = response.body,
                      let parsed = try? JSONDecoder().decode(JSON.self, from: Data(buffer: buffer)),
                      parsed.objectValue != nil else {
                    logger.warning("LLM error    model=\(config.model) elapsed_ms=\(elapsed) error=invalid-json")
                    throw LlmError(message: "LLM response was not valid JSON")
                }
                logger.info("LLM response model=\(config.model) elapsed_ms=\(elapsed) tokens=\(extractTokens(parsed))")
                return parsed
            } catch let e as LlmError {
                throw e
            } catch {
                let elapsed = Int(Date().timeIntervalSince(start) * 1000)
                let willRetry = attempt < attempts
                logger.warning(
                    "LLM error    model=\(config.model) elapsed_ms=\(elapsed) error=\(error)\(willRetry ? " (will retry)" : "")"
                )
                lastError = error
                if willRetry {
                    try? await Task.sleep(nanoseconds: UInt64(config.retryBackoffMs * attempt) * 1_000_000)
                }
            }
        }
        throw LlmError(message: lastError.map { "\($0)" } ?? "LLM request failed")
    }

    private func extractTokens(_ body: JSON) -> String {
        guard let usage = body["usage"]?.objectValue else { return "n/a" }
        if let total = usage["total_tokens"]?.intValue, total > 0 { return String(total) }
        let sum = (usage["input_tokens"]?.intValue ?? 0) + (usage["output_tokens"]?.intValue ?? 0)
        return sum > 0 ? String(sum) : "n/a"
    }

    private func extractChatText(_ body: JSON) throws -> String {
        guard let choices = body["choices"]?.arrayValue,
              let content = choices.first?["message"]?["content"]?.stringValue,
              !content.isEmpty else {
            throw LlmError(message: "LLM chat response missing content")
        }
        return content
    }

    private func extractResponsesText(_ body: JSON) throws -> String {
        if let text = body["output_text"]?.stringValue, !text.isEmpty { return text }
        for item in body["output"]?.arrayValue ?? [] {
            for entry in item["content"]?.arrayValue ?? [] {
                if let text = entry["text"]?.stringValue, !text.isEmpty { return text }
            }
        }
        throw LlmError(message: "LLM response did not contain output text")
    }

    /// 解析 LLM 输出的 JSON 对象：剥代码块围栏；失败时截取首尾花括号再试一次。
    private func parseJsonObject(_ raw: String) throws -> JSON {
        var text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if text.hasPrefix("```") {
            text = text.replacingOccurrences(of: "^```[a-zA-Z]*\\s*", with: "", options: .regularExpression)
                .replacingOccurrences(of: "\\s*```$", with: "", options: .regularExpression)
                .trimmingCharacters(in: .whitespacesAndNewlines)
        }
        if let parsed = try? JSONDecoder().decode(JSON.self, from: Data(text.utf8)),
           parsed.objectValue != nil {
            return parsed
        }
        guard let start = text.firstIndex(of: "{"), let end = text.lastIndex(of: "}"), start < end,
              let parsed = try? JSONDecoder().decode(JSON.self, from: Data(text[start...end].utf8)),
              parsed.objectValue != nil else {
            throw LlmError(message: "LLM output was not valid JSON")
        }
        return parsed
    }

    private var baseUrl: String {
        let trimmed = config.baseUrl.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty { return "https://api.openai.com/v1" }
        return String(trimmed.reversed().drop(while: { $0 == "/" }).reversed())
    }

    private var responsesUrl: String { baseUrl + "/responses" }
    private var chatUrl: String { baseUrl + "/chat/completions" }

    // ── 结构化输出 schema ─────────────────────────────────────────────────

    static let suggestionSpec = SchemaSpec(
        schemaName: "capsule_suggestion",
        schema: .object([
            "type": .string("object"),
            "additionalProperties": .bool(false),
            "required": .array([.string("title"), .string("content"), .string("openInDays")]),
            "properties": .object([
                "title": .object(["type": .string("string")]),
                "content": .object(["type": .string("string")]),
                "openInDays": .object([
                    "type": .string("integer"), "minimum": .int(1), "maximum": .int(3650),
                ]),
            ]),
        ]),
        systemPrompt: "你只返回严格 JSON 对象，不要 Markdown、代码块或解释。"
            + "JSON 必须包含字符串字段 title、content 和整数字段 openInDays。若用户已给出标题，title 可原样回填。",
        maxOutputTokens: 900,
        maxTokens: 900
    )

    static let recommendationSpec = SchemaSpec(
        schemaName: "capsule_recommendations",
        schema: .object([
            "type": .string("object"),
            "additionalProperties": .bool(false),
            "required": .array([.string("items")]),
            "properties": .object([
                "items": .object([
                    "type": .string("array"),
                    "minItems": .int(3),
                    "maxItems": .int(8),
                    "items": .object([
                        "type": .string("object"),
                        "additionalProperties": .bool(false),
                        "required": .array([.string("title"), .string("hint"), .string("openInDays")]),
                        "properties": .object([
                            "title": .object(["type": .string("string")]),
                            "hint": .object(["type": .string("string")]),
                            "openInDays": .object([
                                "type": .string("integer"), "minimum": .int(1), "maximum": .int(3650),
                            ]),
                        ]),
                    ]),
                ]),
            ]),
        ]),
        systemPrompt: "你只返回严格 JSON 对象，不要 Markdown、代码块或解释。"
            + "JSON 必须包含数组字段 items，每项含字符串字段 title、hint 和整数字段 openInDays。",
        maxOutputTokens: 900,
        maxTokens: 900
    )
}
